_drafts/cortex-m-breakpoints.md

```diff
@@ -0,0 +1,607 @@
+
+## Cortex-M Hardware Breakpoints
+
+Now that we have a pretty good idea of how GDB requests breakpoints get set and removed let's see if we can figure out how the gdb server is actually installing them for the NRF52 Cortex-M target.
+
+### Flash Patch & Breakpoint Unit
+
+For ARM Cortex-M MCUs, hardware breakpoint functionality is exposed via the Flash Patch and Breakpoint Unit (**FPB**)[^10].
+
+The FPB has two main features:
+
+1. Allows hardware breakpoints to be configured and enabled.
+2. Allows patching of Flash Code & Data. Historically this has been a popular way to debug ROM bootloaders or ship fixes for devices that can't perform full over the air updates. These days more and more embedded devices perform full firmware updates and so the practice is becomming less common. The art of flash patching deserves a post of its own so we won't be digging deeper here.
+
+For breakpoint management, there are two register types you will care about inside the **FPB**, `FP_CTRL` & `FP_COMPn`.
+
+#### Flash Patch Control Register, FP_CTRL, 0xE0002000
+
+![](/img/breakpoint/fp-ctrl.png)
+
+The `FP_CTRL` is where you can find out information about how many hardware breakpoints are supported and enable the **FPB**. Notably,
+
+- `REV`: There are only two revisions available today, Version 1 (`REV=0`) & Version 2 (`REV=1`). Depending on the revision, the layout of the `FP_COMPn` register will be slightly different. The notable difference between the two versions is that for Version 1 breakpoints can _only_ be installed if the address is in the ARM Code Region range (`0x00000000-0x1FFFFFFF`) but with Version 2 breakpoints can be installed over the entire 4GB address space.
+- Number of Available Hardware Breakpoints can be computed by: `FP_CTRL[6:4] << 4 | FP_CTRL[3:0]`
+- `ENABLE` turns on the **FPB** Unit when set to 1 and disables it when set to 0.
+
+> NOTE: A write to the `FP_CTRL` register will only take effect if the `KEY` bit is 1. Otherwise, the operation will be a no-op.
+
+#### Flash Patch Comparator register, FP_COMPn, 0xE0002008- 0xE0002008+4n
+
+Comparators for breakpoints will always be `FP_COMP0` up to the number of supported breakpoints minus 1. Technically up to 127 hardware breakpoints can be implemented but usually only 2-6 are.
+
+##### Version 1 Layout
+
+![](/img/breakpoint/fp-comp-rev1.png)
+
+`ENABLE` flips on the individual comparator and `REPLACE` controls the behavior as follows:
+
+| Replace | Behavior                                                                         |
+| ------- | -------------------------------------------------------------------------------- |
+| 0b00    | N/A - Used to enable Flash Patching                                              |
+| 0b01    | Install breakpoint on instruction at `'000':COMP:'00'`.                          |
```

Comment:
I don't think this was super explicit that this means:
`0x000XXXXXXXXXXX00` where `X`s are taken from the value of `COMP`. Took me a second to understand.

Reply:
What a dumb design... Can you think of a reason why they would have done it this way?

Reply:
I went with `'000':FP_COMP[28:2]:'00'.`. which I think is more explicit. I got confused initially too :)

Yeah, I can't think of a great reason for the layout. I think maybe someone was trying to be extra clever with the fact that arm has 16bit & 32bit instructions but even that isn't a very strong reason.